content/advent-2015/composable-pipelines-improvements.md

```diff
@@ -0,0 +1,268 @@
++++
+author = ["Samuel Lampa"]
+date = "2015-12-18T00:00:00-08:00"
+title = "Composable Pipelines Improved"
+series = ["Advent 2015"]
++++
+
+I wrote a [post here on GopherAcademy](https://blog.gopheracademy.com/composable-pipelines-pattern) earlier this year,
+about an idea for a "framework-less" pattern for [Flow-Based Programming](http://www.jpaulmorrison.com/fbp/) style programs in Go,
+or let's just call it "composable concurrent pipelines". During the year, I have experimented more, and added some minor modifications to this pattern, which
+I describe below.
+
+## The old way
+
+The basic idea described in that earlier post was the following: Rather to do [like in the generator "pattern"](https://talks.golang.org/2012/concurrency.slide#25),
+where concurrent processes were packed into functions that returns the output channels on which their lazy-evaluated output will be provided,
+we instead store each process into structs, with each in-port and out-port (I use the term port from Flow-based programming here, though technically in Go, it
+is in this case just a channel stored in a named struct field) as separate struct fields, and the main code of the processes in a
+go-routine, to make it concurrent.
+
+The basic approach has the benefits that the code for routing out-ports to in-ports becomes a lot clearer
+than with the generator pattern, since each port is handled separately, even on separate lines of code.
+
+An unnecessary feat of that provided pattern though, was using methods for each out-ports, that would ensure that a channel
+is created for that out-port, and then returning that channel. The pattern proposed that a process would be defined like so:
+
+```go
+type AProcess struct {
+	In  string
+	Out string
+}
+
+func (p *AProcess) OutChan() chan string {
+	p.Out = make(chan []byte, 16)
+	return p.Out
+}
+
+func (p *AProcess) Init() {
+	go func() {
+		for line := range p.In {
+			// Do something with `line` ...
+		}
+	}()
+}
+```
+
+Connecting such processes together would go something like:
+
+```go
+proc1 := &AProcess{}
+proc2 := &AProcess{}
+// Connect
+proc2.In = proc1.OutChan()
+```
+
+## A better way
+
+What I realized, is that if we use a function to create new tasks, that pre-populates all the channel fields, we can
+just assign those fields to each other.
```

Comment:
If you remove the commas and make _pre-populates all_ into _pre-populate_ this would be easier to read. Can you consider updating?